drivers/led/lp50xx.c

```diff
@@ -0,0 +1,518 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ * Copyright (c) 2022 Grinn
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief LP50xx LED controller
+ */
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/drivers/led/lp50xx.h>
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/util.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(lp50xx, CONFIG_LED_LOG_LEVEL);
+
+#define LP50XX_MIN_BRIGHTNESS		0U
+#define LP50XX_MAX_BRIGHTNESS		100U
+
+#define LP5009_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5012_MAX_LEDS)
+#define LP5012_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5012_MAX_LEDS)
+#define LP5018_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5024_MAX_LEDS)
+#define LP5024_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5024_MAX_LEDS)
+#define LP5030_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5036_MAX_LEDS)
+#define LP5036_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5036_MAX_LEDS)
+
+#define LP50XX_ENABLE_DELAY_US		500
+
+/* Common registers */
+#define LP50XX_DEVICE_CONFIG0		0x00
+#define LP50XX_DEVICE_CONFIG1		0x01
+#define LP50XX_LED_CONFIG0		0x02
+
+/* LP5009/12 registers */
+#define LP5012_LED0_BRIGHTNESS		0x07
+#define LP5012_OUT0_COLOR		0x0b
+#define LP5012_RESET			0x17
+
+/* LP5018/24 registers */
+#define LP5024_LED0_BRIGHTNESS		0x07
+#define LP5024_OUT0_COLOR		0x0f
+#define LP5024_RESET			0x27
+
+/* LP5030/36 registers */
+#define LP5036_LED_CONFIG1		0x03
+#define LP5036_LED0_BRIGHTNESS		0x08
+#define LP5036_OUT0_COLOR		0x14
+#define LP5036_RESET			0x38
+
+/* Register values */
+#define CONFIG0_CHIP_EN			BIT(6)
+
+#define CONFIG1_LED_GLOBAL_OFF		BIT(0)
+#define CONFIG1_MAX_CURRENT_OPT		BIT(1)
+#define CONFIG1_PWM_DITHERING_EN	BIT(2)
+#define CONFIG1_AUTO_INCR_EN		BIT(3)
+#define CONFIG1_POWER_SAVE_EN		BIT(4)
+#define CONFIG1_LOG_SCALE_EN		BIT(5)
+
+#define RESET_SW			0xFF
+
+/* Chipset models supported by this driver. */
+enum lp50xx_chips {
+	LP5009,
+	LP5012,
+	LP5018,
+	LP5024,
+	LP5030,
+	LP5036,
+};
+
+struct lp50xx_config {
+	struct i2c_dt_spec bus;
+	const struct gpio_dt_spec gpio_enable;
+	enum lp50xx_chips chip;
+	uint8_t num_leds;
+	bool log_scale_en;
+	bool max_curr_opt;
+	const struct led_info *leds_info;
+};
+
+struct lp50xx_data {
+	uint8_t *chan_buf;
+};
+
+static int lp50xx_get_max_leds(enum lp50xx_chips id) {
+	switch (id) {
+	case LP5009:
+	case LP5012:
+		return LP5012_MAX_LEDS;
+	case LP5018:
+	case LP5024:
+		return LP5024_MAX_LEDS;
+	case LP5030:
+	case LP5036:
+		return LP5036_MAX_LEDS;
+	default:
+		return -EINVAL;
+	}
+
+	return -EINVAL;
```

Comment:
That is superfluous.

Reply:
fixed

Reply:
IMO. it would be better if you removed that line and restore the `default` within the switch case.